src/ophyd_async/core/signal.py

```diff
@@ -63,29 +70,54 @@ def __init__(
     ) -> None:
         self._timeout = timeout
         self._initial_backend = self._backend = backend
+
         super().__init__(name)
 
+    def children(self) -> Iterator[Tuple[str, "Device"]]:
+        yield from ()
+
     async def connect(
         self,
         mock=False,
         timeout=DEFAULT_TIMEOUT,
         force_reconnect: bool = False,
         backend: Optional[SignalBackend[T]] = None,
     ):
-        if backend:
-            if self._initial_backend and backend is not self._initial_backend:
-                raise ValueError(
-                    "Backend at connection different from initialised one."
-                )
-            self._backend = backend
-        if mock and not isinstance(self._backend, MockSignalBackend):
-            # Using a soft backend, look to the initial value
-            self._backend = MockSignalBackend(initial_backend=self._backend)
-
-        if self._backend is None:
-            raise RuntimeError("`connect` called on signal without backend")
-        self.log.debug(f"Connecting to {self.source}")
-        await self._backend.connect(timeout=timeout)
+        """
+        Connect self with a timeout
+
+        Parameters
+        ----------
+        mock : bool, optional
+            uses the `MockSignalBackend`, by default False
+        timeout : float, optional
+            Time to wait before failing with a TimeoutError, by default DEFAULT_TIMEOUT
+        force_reconnect : bool, optional
+            Always force reconnection, by default False
+        """
+        can_use_previous_connection: bool = (
+            self._connect_task
+            and not (self._connect_task.done() and self._connect_task.exception())
+            and self._connect_mock_arg == mock
+        )
+        if force_reconnect or not can_use_previous_connection:
+            self._backend = (
+                MockSignalBackend(initial_backend=self._initial_backend)
+                if mock
+                else backend
+                if backend
+                else self._initial_backend
+            )
```

Comment:
```suggestion
            if backend:
                if self._initial_backend and backend is not self._initial_backend:
                    raise ValueError(
                        "Backend at connection different from initialised one."
                    )
                self._backend = backend
            if mock and not isinstance(self._backend, MockSignalBackend):
                # Using a soft backend, look to the initial value
                self._backend = MockSignalBackend(initial_backend=self._backend)

            if self._backend is None:
                raise RuntimeError("`connect` called on signal without backend")
```

I think this section should be put back to how it currently is - this change

* misses out meaningful error messages
* doesn't skip `MockSignalBackend` initialisation if `initial_backend` is a `MockSignalBackend`
* doesn't do any checking for `initial_backend` vs `backend`

Reply:
that looks quite right, I'm not sure how it ended up changed so much

Reply:
ok I know why - because I was emulating the logic from 'Device'